Add support for running IOMCU FW on CubeRedSecondary

bugobliterator opened the pull request:
Tested working DSHOT, PWM modes and RC Input on CubredSecondary with IOFirmware.

No size changes for older iomcus, checked using sizecompare for both with and without dshot.

Allows updates to the secondary MCU's firmware even after the IOFirmware is flashed.

andyp1per commented on libraries/AP_HAL_ChibiOS/Scheduler.cpp:
```diff
@@ -58,6 +58,14 @@
 extern AP_IOMCU iomcu;
 #endif
 
+#ifdef IOMCU_FW
```
This is only with H7 I think? Probably should add that to the define.

bugobliterator replied:
it gets removed for other boards that don't use it by compiler. As shown by no size difference.

andyp1per commented on libraries/AP_IOMCU/AP_IOMCU.cpp:
```diff
@@ -100,6 +104,34 @@ void AP_IOMCU::init(void)
 }
 
+/*
+  disable the IOMCU operation by erasing iofirmware
```
I think this should have a different name given the fairly radical function of its operation

bugobliterator replied:
thoughts on name, erase is already taken.

bugobliterator replied:
changing it to wipe

bugobliterator replied:
removed completely

andyp1per commented on libraries/AP_IOMCU/fw_uploader.cpp:
```diff
@@ -281,7 +290,7 @@ bool AP_IOMCU::erase()
     send(PROTO_CHIP_ERASE);
     send(PROTO_EOC);
-    return get_sync(10000);
+    return get_sync(20000);
```
Is this a timeout? Maybe add a comment

bugobliterator replied:
done

andyp1per commented on libraries/AP_IOMCU/iofirmware/analog.cpp:
```diff
@@ -157,3 +159,54 @@ OSAL_IRQ_HANDLER(STM32_ADC1_HANDLER) {
+}
+
+/*
+  capture VSERVO in mV
```
Comments are wrong. Should add a comment as to why these are empty

bugobliterator replied:
fixed

andyp1per commented on libraries/AP_IOMCU/iofirmware/iofirmware.cpp:
```diff
@@ -73,9 +103,17 @@ static void setup_rx_dma(hal_uart_driver* uart)
-                     STM32_DMA_CR_MINC | STM32_DMA_CR_TCIE);
+                     STM32_DMA_CR_MINC | STM32_DMA_CR_TCIE
+#if defined(STM32H7)
+                     | (1<<20)
```
needs a define

bugobliterator replied:
fixed through define and also in UARTDriver

andyp1per commented on libraries/AP_IOMCU/iofirmware/iofirmware.cpp:
```diff
@@ -87,11 +125,23 @@ static void setup_tx_dma(hal_uart_driver* uart)
-                     STM32_DMA_CR_MINC | STM32_DMA_CR_TCIE);
+                     STM32_DMA_CR_MINC | STM32_DMA_CR_TCIE
+#if defined(STM32H7)
+                     | (1<<20)
```
ditto

bugobliterator replied:
fixed

andyp1per commented on libraries/AP_IOMCU/iofirmware/iofirmware.cpp:
```diff
@@ -147,22 +207,43 @@ static void dma_tx_end_cb(hal_uart_driver *uart)
-        uart->usart->CR1 = cr1 | USART_CR1_SBK;
-
+        CLEAR_ISR_FLAG(LBD);
+        SEND_BREAK();
```
this is a slight change in behaviour because this was a local variable for threading reasons. Ideally would preserve the current semantics

bugobliterator replied:
I am going to change it back to original as it doesn't really matter in this context.

But the change is actually correct. Firstly a thread cannot interrupt an interrupt, another higher priority interrupt can. Secondly, only bit we want to set in CR1 is Line Break, rest of the bits we don't want to touch. With the older method there's a chance we can end up doing just that.
Basically if there was another interrupt or peripheral event (between where we store cr1 and call SEND_BREAK) where the CR1 is changed we will end up changing it again unknowingly. Fortuntely we don't mess with CR1 outside of init and this interrupt, so it is a non-issue. But in general this can cause issue.

andyp1per commented on libraries/AP_IOMCU/iofirmware/iofirmware.cpp:
```diff
@@ -176,7 +257,7 @@ static void idle_rx_handler(hal_uart_driver *uart)
     if ((sr & USART_SR_TC) && (cr1 & USART_CR1_TCIE)) {
         /* TC interrupt cleared and disabled.*/
-        uart->usart->SR &= ~USART_SR_TC;
+        CLEAR_ISR_FLAG(TC);
```
Might be better to use the MODIFY_REG() macro that you used in bdshot

andyp1per replied:
ok

andyp1per commented on libraries/AP_IOMCU/iofirmware/iofirmware.cpp:
```diff
@@ -214,11 +298,11 @@ static void uart_lld_serve_tx_end_irq(hal_uart_driver *uart, uint32_t flags)
-        uart->dmatx = dmaStreamAllocI(STM32_UART_USART2_TX_DMA_STREAM,
-                                        STM32_UART_USART2_IRQ_PRIORITY,
+        uart->dmatx = dmaStreamAllocI(HAL_IO_FMU_COMMS_TX_DMA_STREAM,
+                                        12,  //IRQ Priority
```
This is bypassing the config, thing it should use the define or put a define above

bugobliterator replied:
yeah, I was trying to avoid adding yet another define that's going to remain untouched. Will add this to define as well.

tridge replied:
still pending?

andyp1per commented on libraries/AP_IOMCU/iofirmware/iofirmware.cpp:
```diff
@@ -447,7 +567,7 @@ void AP_IOMCU_FW::update()
 
     if (do_reboot && (last_ms > reboot_time)) {
-        hal.scheduler->reboot(true);
+        hal.scheduler->reboot(false);
```
why?

andyp1per replied:
ok sounds reasonable. Probably needs testing on other boards that have iofirmware update issues - particularly Pixhawk6X

andyp1per reviewed (changes requested):
Looking good. Some small changes. Think you need to think about how it looks if someone adds F4 support (which has been requested)

andyp1per reviewed (approved):
LGTM, just needs testing iofirmware upgrades on some of the other boards (e.g. Pixhawk6X)

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:
```diff
@@ -1903,11 +1914,19 @@ def write_UART_config(self, f):
-                '#define HAL_UART_IO_DRIVER ChibiOS::UARTDriver uart_io(HAL_UART_IOMCU_IDX)\n'
-            )
-            serial_list.append(self.config['IOMCU_UART'][0])
+            if self.config['IOMCU_UART'][0].startswith('HAL_SERIAL'):
```
if we made AP_IOMCU take a pointer to a AP_HAL::UARTDriver instead of a reference then we could have an AP::iomcu()->set_uart() call that could be made from AP_SerialManager. That would enable runtime configuration.

bugobliterator replied:
I think I am going to do this in a separate PR, as that will require quite a decent amount of change to AP_IOMCU. A separate PR for this change will be much easier to review.

For this PR I can go part of the way and use SerialProtocol and hal.serial instead of current method of declairing a separate serial driver for all iomcu users.

tridge commented on libraries/AP_Vehicle/AP_Vehicle.cpp:
```diff
@@ -356,6 +356,15 @@ void AP_Vehicle::setup()
+    if (BoardConfig.io_enabled()) {
+        // If IO is enabled, we need to set the protocol to None
+        // as the IO MCU UART is handled by the AP_IOMCU class
+        serial_manager.set_protocol_and_baud(HAL_UART_IOMCU_IDX, AP_SerialManager::SerialProtocol_None, 0);
```
SerialProtocol_IOMCU would make more sense

bugobliterator replied:
changed to SerialProtocol_IOMCU

tridge commented on libraries/AP_Vehicle/AP_Vehicle.cpp:
```diff
@@ -356,6 +356,15 @@ void AP_Vehicle::setup()
+        // as the IO MCU UART is handled by the AP_IOMCU class
+        serial_manager.set_protocol_and_baud(HAL_UART_IOMCU_IDX, AP_SerialManager::SerialProtocol_None, 0);
+    } else {
+        iomcu.wipe();
```
why are we wiping the fw if IO not enabled?

bugobliterator replied:
This is to allow flashing a different firmware to secondary. We have to reduce the boot time if IOMCU is enabled to 200ms, which is not long enough for it to be caught over USB. To allow updating Secondary MCU over USB again we need wipe iofirmware.

bugobliterator replied:
Not relevant anymore, moved to different procedure to allow updates other than iofirmware.

tridge reviewed (changes requested):
I'm glad to see this becoming an option, just need to work out how to properly dynamically choose if UART is IOMCU or some other protocol, the method in this PR doesn't look right
also, this does change the size of iofirmware_f103_lowpolh.bin by 8 bytes, need to work out why

bugobliterator commented on libraries/AP_IOMCU/iofirmware/analog.cpp:
```diff
@@ -22,6 +22,8 @@
 #include "hal.h"
 #include "analog.h"
 
+#ifdef HAL_DISABLE_ADC_DRIVER
```
no, that is correct. If AP::HAL ADC Driver is disabled analog.cpp handles ADC periph

Otherwise we use hal.analogin

bugobliterator commented on Tools/AP_Bootloader/support.cpp:
```diff
@@ -240,6 +240,17 @@ bool flash_write_buffer(uint32_t address, const uint32_t *v, uint8_t nwords)
+ */
+#if defined(HAL_BL_IOMCU_FW_DETECT) && HAL_BL_IOMCU_FW_DETECT
+#define HAL_IOMCU_MAGIC 0x937E638A
+uint32_t is_iomcu_fw_present(void)
```
This is so we can reduce the boot time if IOMCU is present to 200ms,. This is done by reading the IOMCU_FW_DETECT magic written to the end of flash. 

I think I am going to add a boot time field to AP_CheckFirmware instead. Was looking for a shortcut but this hasn't turned out to be a clean implementation.

bugobliterator replied:
not relevant anymore, this has been removed now.

bugobliterator commented on Tools/AP_Bootloader/AP_Bootloader.cpp:
```diff
@@ -87,6 +87,12 @@ int main(void)
     uint32_t timeout = HAL_BOOTLOADER_TIMEOUT;
 
+#if defined(HAL_BL_IOMCU_FW_DETECT) && HAL_BL_IOMCU_FW_DETECT
+    if (is_iomcu_fw_present()) {
```
Current iofirmware bootloader has 200ms timeout. Also if iofirmware doesn't boot faster than when AP_IOMCU tries to detect it, it will simply try to reflash.

bugobliterator commented on libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:
```diff
@@ -1052,6 +1052,17 @@ def write_mcu_config(self, f):
+            self.env_vars['IOMCU_FW'] = 0
+
+        # check if heater pin defined
+        if 'HEATER' in self.bylabel.keys():
```
This is needed to disable multiple builds for highpol and lowpol heaters. With hwdef without Heaters we were still doing that.

bugobliterator commented:
@tridge I have pushed changes per your comments.

* got rid of the bootloader changes, and using bootloader CRC to verify IOMCU at boot.
* Added option SerialProtocol_IOMCU, currently serialmanager is not doing anything with it.
* moved to using serialxDriver for IOMCU as well.

bugobliterator commented on libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:
```diff
@@ -1052,6 +1052,17 @@ def write_mcu_config(self, f):
         else:
             self.env_vars['IOMCU_FW'] = 0
 
+        if self.get_config('IOMCU_FW', required=False):
```
fixed

peterbarker commented on libraries/AP_HAL_ChibiOS/hwdef/CubeRedSecondary-IO/hwdef.dat:
```diff
@@ -0,0 +1,170 @@
+
+define HAL_USE_RTC FALSE
+define HAL_NO_UARTDRIVER TRUE
+define HAL_LOGGING_ENABLED 0
```
Why are these in here?  These are the default for io firmware... (defaults_periph.h)

bugobliterator replied:
They are not actually, only HAL_LOGGING_ENABLED is. In anycase, I copied this from iomcu/hwdef.inc

peterbarker replied:
> They are not actually, only HAL_LOGGING_ENABLED is. In anycase, I copied this from iomcu/hwdef.inc

`HAL_NO_UARTDRIVER` no longer exists.

If `HAL_USE_RTC` needs a default to "off" for IOMCU (and it does!) that default should go into `defaults_iofirmware.h`.  The HAL_LOGGING_ENABLED in `iomcu/hwdef.inc` is redundant as it already appears in `defaults_iofirmware.h`

peterbarker commented on libraries/AP_HAL_ChibiOS/hwdef/CubeRedSecondary-IO/hwdef.dat:
```diff
@@ -0,0 +1,170 @@
+define HAL_GPIO_PIN_SPEKTRUM_OUT PAL_LINE(GPIOC,11U)
+
+
+define AP_NETWORKING_BACKEND_PPP 1
```
This is odd?

bugobliterator replied:
fixed

peterbarker commented on libraries/AP_IOMCU/iofirmware/rc.cpp:
```diff
@@ -27,27 +27,57 @@
-static uint8_t sd3_config;
+static uint8_t sbus_out_config;
+
+#define STR(x) XSTR(x)
+#define XSTR(x) #x
+
+
```
Unused?

bugobliterator replied:
fixed

tridge commented on libraries/AP_IOMCU/fw_uploader.cpp:
```diff
@@ -417,12 +429,13 @@ bool AP_IOMCU::verify_rev3(uint32_t fw_size_local)
     crc_is_ok = true;
 
-    return true;
+    return get_sync();
```
ifdef

bugobliterator replied:
done

bugobliterator replied:
fixed

peterbarker commented on libraries/AP_HAL_ChibiOS/Scheduler.cpp:
```diff
@@ -95,6 +95,14 @@ THD_WORKING_AREA(_monitor_thread_wa, MONITOR_THD_WA_SIZE);
 #define AP_HAL_CHIBIOS_IN_EXPECTED_DELAY_WHEN_NOT_INITIALISED 1
 #endif
 
+#if defined(STM32H7) && defined(IOMCU_FW)
```
```suggestion
#if AP_REBOOT_ON_MEMGUARD_ERROR
```
... and default it off and define to true in the hwdef.

bugobliterator replied:
I think that can be a bit deceptive If I do it the way you mentioned. 

As the normal AP_memory_guard_error is defined inside AP_InternalError. This is actually a special case for IOMCU_FW running  on STM32H7.

bugobliterator commented:
> i'd like to see binary cmp showing no change in IO fw builds

@tridge I did a binary comparision for both iomcu and iomcu-dshot firmware. The difference is only a single byte (1 -> 0) , which is related to a simple change of calling reboot with argument false instead of previously true here:

https://github.com/ArduPilot/ardupilot/pull/29094/files#diff-31df3802edd794b59d66393f1064d74590190327e6423b2ab743e97067afe4f6R102

bugobliterator commented on libraries/AP_IOMCU/fw_uploader.cpp:
```diff
@@ -281,7 +293,7 @@ bool AP_IOMCU::erase()
     send(PROTO_CHIP_ERASE);
     send(PROTO_EOC);
-    return get_sync(10000);
+    return get_sync(20000); // 20s timeout for erase
```
done

bugobliterator replied:
fixed

bugobliterator commented on libraries/AP_IOMCU/fw_uploader.cpp:
```diff
@@ -417,12 +429,13 @@ bool AP_IOMCU::verify_rev3(uint32_t fw_size_local)
     /* compare the CRC sum from the IO with the one calculated */
     if (sum != crc) {
         debug("CRC wrong: received: 0x%x, expected: 0x%x", (unsigned)crc, (unsigned)sum);
+        get_sync();
```
fixed

bugobliterator commented on libraries/AP_IOMCU/fw_uploader.cpp:
```diff
@@ -84,11 +88,11 @@ bool AP_IOMCU::upload_fw(void)
-        debug("Err: failed to contact bootloader");
+    uint32_t bl_rev, unused;
+    if (!get_info(INFO_BL_REV, bl_rev) ||
+        !get_info(INFO_BOARD_ID, unused) ||
```
done

peterbarker commented on libraries/AP_HAL_ChibiOS/hwdef/CubeRedSecondary-IO/hwdef.dat:
```diff
@@ -0,0 +1,166 @@
+define HAL_IOMCU_SBUS_OUT_SERIAL_DRIVER SD6
+define HAL_GPIO_PIN_SPEKTRUM_OUT PAL_LINE(GPIOC,11U)
+
+# only use pulse input for PPM, other protocols
+# are on serial
+define HAL_RCIN_PULSE_PPM_ONLY
+
+
```
This is unused.

Perhaps it would be better to create a `CubeRedSecondary/hwdef.inc` so we're not making existing problems worse?

Note this would be the first time sharing hwdef information between iomcu and a normal board, so it might not work.

bugobliterator replied:
sharing hwdef.inc between IO and Secondary would be quite complex actually, except for pins almost everything will need to be disabled or changed.

peterbarker commented on libraries/AP_HAL_ChibiOS/hwdef/common/common_mixf.ld:
```diff
@@ -74,6 +74,7 @@ SECTIONS
         *(.ap_romfs*)
         /* moving GCS_MAVLink to ext_flash */
         lib/lib*.a:GCS*.*(.text* .rodata*)
+        lib/lib*.a:tables.*(.text* .rodata*) /* tables for AP_Declination */
```
We already have patches for this in master - they're in the libraries themselves rather than hidden in the ld scripts

bugobliterator replied:
can you please share the link to the PRs?

peterbarker replied:
Showed Sid the code on DevCallEU

peterbarker commented on libraries/AP_IOMCU/AP_IOMCU.cpp:
```diff
@@ -1092,10 +1096,14 @@ void AP_IOMCU::send_rc_protocols()
 /*
   check ROMFS firmware against CRC on IOMCU, and if incorrect then upload new firmware
  */
+#ifndef HAL_IOMCU_FW_FLASH_SIZE
```
It would obviously be nicer to always get this from the build / embedding system

Not really needed for this PR.

bugobliterator replied:
ok, will defer it for another time.

peterbarker commented on libraries/AP_IOMCU/fw_uploader.cpp:
```diff
@@ -60,13 +60,21 @@
 {
     // set baudrate for bootloader
-    uart.begin(115200, 256, 256);
+    uart.begin(HAL_IOMCU_BOOTLOADER_BAUDRATE, 256, 256);
```
```suggestion
    uart.begin(AP_IOMCU_BOOTLOADER_BAUDRATE, 256, 256);
```

... similarly for other defines introduced.  `AP_IOMCU` isn't part of the HAL.

bugobliterator replied:
fixed

peterbarker commented on libraries/AP_IOMCU/fw_uploader.cpp:
```diff
@@ -281,7 +307,11 @@ bool AP_IOMCU::erase()
     send(PROTO_CHIP_ERASE);
     send(PROTO_EOC);
-    return get_sync(10000);
+#if HAL_IOMCU_CHIBIOS_BOOTLOADER
+    return get_sync(20000); // 20s timeout for erase
+#else
+    return get_sync(10000); // 10s timeout for erase
+#endif
```
Does the erase take longer just because of `HAL_IOMCU_CHIBIOS_BOOTLOADER`?

Should this instead be `HAL_IOMCU_FW_FLASH_SIZE > (0x10000 - 0x1000)`?

bugobliterator replied:
the reason is flash size, so I guess it can be changes to be based on that.

bugobliterator replied:
fixed

peterbarker commented on libraries/AP_Vehicle/AP_Vehicle.cpp:
```diff
@@ -356,6 +356,13 @@ void AP_Vehicle::setup()
 #if AP_SERIALMANAGER_ENABLED
+#if HAL_WITH_IO_MCU
+    if (BoardConfig.io_enabled()) {
+        // If IO is enabled, we need to set the protocol to None
+        // as the IO MCU UART is handled by the AP_IOMCU class
```
What happens if we *don't* do this?

bugobliterator replied:
there's a typo that it needs to say IOMCU not None

bugobliterator replied:
comment removed altogether, as its pretty clear from the code what is happening.
